dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -0,0 +1,182 @@
+#include <winerror.h>
+#include <algorithm>
+#include <iostream>
+#include "PushNotificationChannel.h"
+#include <wil/resource.h>
+
+using namespace winrt::Windows::Networking::PushNotifications;
+using namespace winrt;
+
+namespace winrt::Microsoft::Windows::PushNotifications::implementation
+{
+    inline constexpr std::uint32_t c_maxBackoffSeconds{ 960 };
+    inline constexpr std::uint32_t c_minBackoffSeconds{ 30 };
+
+    const HRESULT WNP_E_NOT_CONNECTED = MAKE_HRESULT(SEVERITY_ERROR, FACILITY_ITF, 0x880403E8L);
+    const HRESULT WNP_E_RECONNECTING = MAKE_HRESULT(SEVERITY_ERROR, FACILITY_ITF, 0x880403E9L);
+    const HRESULT WNP_E_BIND_USER_BUSY = MAKE_HRESULT(SEVERITY_ERROR, FACILITY_ITF, 0x880403FEL);
+
+    bool PushNotificationManager::isChannelRequestRetryable(const hresult& hr)
+    {
+        switch (hr)
+        {
+        case HRESULT_FROM_WIN32(ERROR_TIMEOUT):
+        case WNP_E_NOT_CONNECTED:
+        case WPN_E_OUTSTANDING_CHANNEL_REQUEST:
+        case WNP_E_RECONNECTING:
+        case WNP_E_BIND_USER_BUSY:
+        case RPC_S_SERVER_UNAVAILABLE:
+            return true;
+        default:
+            return false;
+        }
+    }
+
+    inline bool IsPackagedProcess()
+    {
+        wchar_t packageFullName[PACKAGE_FULL_NAME_MAX_LENGTH + 1] = {};
+
+        UINT32 packageFullNameLength = static_cast<UINT32>(ARRAYSIZE(packageFullName));
+
+        const auto packagedProcessHResult = ::GetCurrentPackageFullName(&packageFullNameLength, packageFullName);
+
+        if (packagedProcessHResult == APPMODEL_ERROR_NO_PACKAGE)
+        {
+            return false;
+        }
+        else if(packagedProcessHResult != S_OK)
+        {
+            winrt::throw_hresult(packagedProcessHResult);
+        }
+
+        return true;
+    }
+
+    winrt::Windows::Foundation::IAsyncOperationWithProgress<Microsoft::Windows::PushNotifications::PushNotificationCreateChannelResult, Microsoft::Windows::PushNotifications::PushNotificationCreateChannelStatus> PushNotificationManager::CreateChannelAsync(winrt::guid remoteId)
+    {
+        static bool s_remoteIdInProgress;
+        static wil::critical_section s_lock;
+
+        if (remoteId == winrt::guid())
+        {
+            throw_hresult(E_INVALIDARG);
+        }
+
+        // API supports channel requests only for packaged applications for v0.8 version
+        bool isPackagedProcess = IsPackagedProcess();
+
+        if (!isPackagedProcess)
+        {
+            throw_hresult(E_NOTIMPL);
+        }
+
+        bool remoteIdInProgress = false;
+        {
+            auto lock = s_lock.lock();
+            if (s_remoteIdInProgress == false)
+            {
+                s_remoteIdInProgress = (!remoteIdInProgress);
+            }
+            else
+            {
+                remoteIdInProgress = true;
+            }
+        }
+
+        winrt::Microsoft::Windows::PushNotifications::PushNotificationCreateChannelResult channelResult{ nullptr };
```

Comment:
Is there a way to make a nullptr constructor with a no throw guarantee?

Reply:
there is nothing being allocated here for a throw to arise. It is just setting a variable to nullptr. that is it. Memory for it is being allocated at a later point. Thanks.